article.md

```diff
@@ -0,0 +1,322 @@
+    version: '1',
+  },
+  message: {
+    "starknetAddress": 12356789,
+  },
+  primaryType: 'Payload',
+  types: {
+    domain: [
+      { name: 'name', type: 'string' },
+      { name: 'verifyingContract', type: 'address' },
+      { name: 'version', type: 'string' },
+    ],
+    Payload: [
+      { name: 'starknetAddress', type: 'uint256' },
+    ],
+  },
+};
+const accounts = await window.ethereum.request({method: "eth_requestAccounts"})
+await window.ethereum.request({
+    method: "eth_signTypedData_v4",
+    from: accounts[0],
+    params: [accounts[0], JSON.stringify(msg)]
+})
+```
+After you run this this JS code in the browser's console on the MetaMask website it will trigger a signing request. An application would use a similar code to get user's accounts and signature. <br/><br/>
+<img alt="Screenshot 2022-10-12 at 10 41 54" src="./graphics/metamask.png"><br/><br/>
+As you can see everything is transparent so you don't have to sign something blindly - that's one of the pros of using EIP-712 standard.<br/><br/>
+<img alt="Screenshot 2022-10-12 at 10 41 54" src="./graphics/signature.png"><br/>
+When you sign this message you will receive the message signature that we need. After that we can send a transaction containing both the Ethereum address and the signature to StarkNet's contract.
+## Creating a connection
+Before we create a connection we need to ensure that the caller owns both the addresses. This is how we can do it:
+1. Sign the structure containing the StarkNet address with our Ethereum account
+2. Call the contract that we are creating with received signature and our Ethereum address from your StarkNet account.
+3. By calling `get_caller_address()` and using it as our Starknet address we ensure that it indeed belongs to the user and we will use it when reconstructing a signed message.
+4. Recreate the message structure and hash it with keccak256.
+5. Extract Ethereum address from Ethereum signature and the hash from previous step.
+6. Compare Ethereum address received as an input to the address extracted from the step before.
+
+The hardest part for us is getting the hash of our message. Once we have that, we can use a built-in function `verify_eth_signature_uint256` that combines steps 5 and 6 and will do all the heavy lifting, dealing with complicated math. 
+So we will have to write a function that constructs an eip712 structure and returns its hash.
+
+## Recreating and hashing the message
+### The basic idea
+Since the ownership of the Ethereum account is confirmed by signing the message for which the private key is needed, we also need to confirm that the user owns the StarkNet account that is supposed to be connected. That's why we're deliberately not passing `starknet_address` as an input variable, instead, we call `get_caller_address()`, which we'll use as the address we need. The connection will be created only if the address of the account executing the tx is the one signed.
```

Comment:
"That's why we can't accept StarkNet address as a parameter and have to use syscall `get_caller_address` to get it."